### ISSUE TITLE
Sharing and distributing log files/build metadata?

### ISSUE THREAD

**Issue by thoughtpolice** [facebook/buck2#441]
`buck2 log` is pretty cool, since it lets you look at what build commands a user ran and what their output is, and you can do things like `log cmd` or `log replay` to watch the build. Is it possible or recommended to 'share' these log files? What would some infrastructure for doing that look like? Maybe something like:

- Users run `buck2 build ...` a bunch
- Their logs can get synchronized to something like an HTTP endpoint (maybe a proxy that just fronts an S3 bucket)
- Something goes wrong, a user asks for help, posts their Build ID.
- You could run `buck2 log cmd --trace-id $BUILD_ID` to find out what happened.
- This would (transparently?) download the log files from some endpoint (somehow?)

There could be other useful things to derive from this possibly, assuming you had these logs. For example, you could use these logs as a direct source of build analytic information from users to derive information about build times, etc.

Is there any kind of "thing" like this inside Meta? Does this seem interesting? To be truly useful it would need some support in the core executable, I think. It seems like something *kind of* like what `buck2 rage` does, right? Except I'd want it in all cases, not just failures.

I imagine this could look like the following for OSS users:

- A `.buckconfig` key like `buck2_logs.upload_address` is set to `https://buck2-logs.aseipp.dev/`
- All `.pb.zst` files are POST'd to `$UPLOAD_ADDRESS/upload` with their trace ID as a primary key, asynchronously, by the daemon
- When a user says `buck2 log cmd --trace-id $TRACE_ID`, then:
  - HTTP GET `$UPLOAD_ADDRESS/trace/$TRACE_ID`
  - This should return a json object containing metadata and a new address to find the log content at
  - That new address is the public URL to download the `.pb.zst` file
  - This allows users to materialize logs on demand
- Something something something authorization with a bearer token
- Ideally the HTTP protocol is simple enough to implement "by hand" in a short afternoon

**Comment by thoughtpolice** [comment]
Actually, if `--trace-id` could take an `https://` URL and download a file and cache it, you could probably do the first part completely outside of the core executable with a file watching tool? Just watch `buck-out/v2/log` and upload every file that gets written?

**Comment by cjhopman** [comment]
Yeah, I've been really happy with the entire suite of `buck2 log` commands.

> Is there any kind of "thing" like this inside Meta? Does this seem interesting? 

Yes, there is. You can see some stuff related to log uploading here: https://github.com/facebook/buck2/blob/062f014ddfeddbef12542b3e233e2f609eacca9a/app/buck2_client_ctx/src/manifold.rs#L131 and then all the log commands have support for fetching the trace-id from where they are stored around here: https://github.com/facebook/buck2/blob/062f014ddfeddbef12542b3e233e2f609eacca9a/app/buck2_client/src/commands/log/options.rs#L90C1-L112

This has been incredibly useful for understanding and providing support for user builds.

I think it would be great if we could figure out the appropriate API for supporting this for open source users as well. I think it probably should be considered together with https://github.com/facebook/buck2/issues/226. The answer there may be that it's actually best to have separate apis for them (you'll note that internally we are doing them basically entirely separately).

One important thing we've found is that we want to upload the log incrementally as the build progresses to ensure that we get logs even on failed things (particularly our CI may aggressively kill off timing out or ooming things that makes it hard to reliably capture on failures).

**Comment by arianvp** [comment]
I don't feel like the uses of append in the codebase require the need of append-style buckets? In spirit all the calls-sites really seem to be using this as a multi-part upload api?  Couldn't we instead add an explicit `finish()`call  that calls  S3's `FinishMultipartUpload` and make sure all buck2 calls finish the uploads that they start?  the `finish()` method could simply be a no-op for Manifold. 

Or are there truly code-paths that never end up in a `finish()` condition?

**Comment by lf-** [comment]
Doesn't multi part upload in s3 require knowing the final length (or maybe hash) up front? That was my impression which was why I didn't consider it, but maybe I was wrong. 

Also CI jobs get terminated regularly and I would absolutely hate to lose logs for those.

**Comment by arianvp** [comment]
No. Multipart uploads do not require the length upfront. 

The concern about a crashed job (e.g. OOM) not calling the finish method seems legitimate.  Bit of a headscratcher yes.

You need to decide to then call CompleteMultiPartUpload out of band to recover the partially uploaded parts. But how do you decide that you can actually call it?  🤔

**Comment by arianvp** [comment]
Potential idea: have your CI build record the invocation ID immediately at the start of the build. If the build crashes; a follow-up step can:

1. Call `list-multpart-uploads` for the invocation ID to get the upload id
2. Call `list-parts` to list all the so far uploaded parts
3. Call `complete-multipart-uploads`  with all the parts

It's not super elegant but it does work. Not as satisfying as uploading the parts as they come though. Must admit.

Doesn't solve the issue for local logs either. Though I think you can trigger reupload of local logs with `rage` right?

**Comment by lf-** [comment]
fwiw it *is* the case that log upload is out of process so it's going to be mostly safe from crashes, but it still needs to have a clean shutdown when ci kills it. That is probably a tractable problem, but we would have to figure out how to cleanly stop it from out of band without too much awareness of buck2 (i.e. determine the correct signal to send it).
